x-pack/plugin/enrich/src/main/java/org/elasticsearch/xpack/enrich/action/ShardMultiSearchAction.java

```diff
@@ -0,0 +1,244 @@
+        }
+    }
+
+    public static class TransportAction extends TransportSingleShardAction<Request, MultiSearchResponse> {
+
+        private final IndicesService indicesService;
+
+        @Inject
+        public TransportAction(ThreadPool threadPool, ClusterService clusterService, TransportService transportService,
+                               ActionFilters actionFilters, IndexNameExpressionResolver indexNameExpressionResolver,
+                               IndicesService indicesService) {
+            super(NAME, threadPool, clusterService, transportService, actionFilters, indexNameExpressionResolver,
+                Request::new, ThreadPool.Names.SEARCH);
+            this.indicesService = indicesService;
+        }
+
+        @Override
+        protected Writeable.Reader<MultiSearchResponse> getResponseReader() {
+            return MultiSearchResponse::new;
+        }
+
+        @Override
+        protected boolean resolveIndex(Request request) {
+            return true;
+        }
+
+        @Override
+        protected ShardsIterator shards(ClusterState state, InternalRequest request) {
+            String index = request.concreteIndex();
+            IndexRoutingTable indexRouting = state.routingTable().index(index);
+            int numShards = indexRouting.shards().size();
+            if (numShards != 1) {
+                throw new IllegalStateException("index [" + index + "] should have 1 shard, but has " + numShards + " shards");
+            }
+            return indexRouting.shard(0).shardsRandomIt();
+        }
+
+        @Override
+        protected MultiSearchResponse shardOperation(Request request, ShardId shardId) throws IOException {
+            final long nowInMillis = System.currentTimeMillis();
+            final IndexService indexService = indicesService.indexService(shardId.getIndex());
+            final IndexShard indexShard = indicesService.getShardOrNull(shardId);
+            try (Engine.Searcher searcher = indexShard.acquireSearcher("enrich_msearch")) {
+                final FieldsVisitor visitor = new FieldsVisitor(true);
+                final QueryShardContext context =
+                    indexService.newQueryShardContext(shardId.id(), searcher.reader(), () -> nowInMillis, null);
+                final MapperService mapperService = context.getMapperService();
+                final Text typeText = mapperService.documentMapper().typeText();
+
+                final MultiSearchResponse.Item[] items = new MultiSearchResponse.Item[request.multiSearchRequest.requests().size()];
+                for (int i = 0; i < request.multiSearchRequest.requests().size(); i++) {
+                    final SearchSourceBuilder searchSourceBuilder = request.multiSearchRequest.requests().get(i).source();
+
+                    final QueryBuilder queryBuilder = searchSourceBuilder.query();
+                    final int from = searchSourceBuilder.from() == -1 ? 0 : searchSourceBuilder.from();
```

Comment:
Don't we have complete control over the search here ? If so, can we remove this validation (and possible some of the topDocs below) in favor of some documented assumptions and asserts to help keep the assumptions valid.

Reply:
Pushed: 182bcd1e3bfbb86c1434380beb5b939559f67167

I removed the normalization of from and size if not specified and moved
the assertion of the search source builder to the request class. I like
to keep this validation, because this is the easiest way to figure out if
a search request is using an unsupported search feature. In the new place
we fail early; on the calling side.